tools/jdbc/src/jni/duckdb_java.cpp

```diff
@@ -243,23 +243,43 @@ static jobject decode_charbuffer_to_jstring(JNIEnv *env, const char *d_str, idx_
 	return j_str;
 }
 
+/**
+ * Associates a duckdb::Connection with a duckdb::DuckDB. The DB may be shared amongst many ConnectionHolders, but the
+ * Connection is unique to this holder. Every Java DuckDBConnection has exactly 1 of these holders, and they are never
+ * shared. The holder is freed when the DuckDBConnection is closed. When the last holder sharing a DuckDB is freed, the
+ * DuckDB is released as well.
+ */
+struct ConnectionHolder {
+	const shared_ptr<duckdb::DuckDB> db;
+	const unique_ptr<duckdb::Connection> connection;
+
+	ConnectionHolder(shared_ptr<duckdb::DuckDB> _db) : db(_db), connection(make_unique<duckdb::Connection>(*_db)) {}
+};
+
+/**
+ * Throws a SQLException and returns nullptr if a valid Connection can't be retrieved from the buffer.
+ */
 static Connection *get_connection(JNIEnv *env, jobject conn_ref_buf) {
-	if (conn_ref_buf) {
-		auto conn_ref = (Connection *)env->GetDirectBufferAddress(conn_ref_buf);
-		if (conn_ref && conn_ref->context) {
-			return conn_ref;
-		}
+	if (!conn_ref_buf) {
+		env->ThrowNew(J_SQLException, "Invalid connection");
+		return nullptr;
+	}
+	auto conn_holder = (ConnectionHolder *)env->GetDirectBufferAddress(conn_ref_buf);
+	if (!conn_holder) {
+		env->ThrowNew(J_SQLException, "Invalid connection");
+		return nullptr;
+	}
+	auto conn_ref = conn_holder->connection.get();
+	if (!conn_ref || !conn_ref->context) {
+		env->ThrowNew(J_SQLException, "Invalid connection");
+		return nullptr;
 	}
 
-	env->ThrowNew(J_SQLException, "Invalid connection");
-	return nullptr;
+	return conn_ref;
 }
 
 //! The database instance cache, used so that multiple connections to the same file point to the same database object
 duckdb::DBInstanceCache instance_cache;
-std::mutex db_map_lock;
-//! A map of pointer to shared_ptr, to ensure we keep the DuckDB object alive
-std::unordered_map<duckdb::DuckDB *, shared_ptr<duckdb::DuckDB>> db_map;
```

Comment:
This mapping was broken. Yes, it held the `shared_ptr` to avoid automatic cleanup, but it also needed to hold a counter of how many `DuckDBDatabase`s were using the mapping, something that was not in place. Implementing such a counter begins to look a lot like a custom shared_ptr 😜 Instead, the 1 to 1 mapping between `ConnectionHolder` and `DuckDBConnection` achieves the right logic, but using regular C++ code/mechanics.

Reply:
Might this fix https://github.com/duckdb/duckdb/issues/6484 then?